Compiler warnings

Globally on my environment, I set `CXXFLAGS` to `-Wall -Wextra -Wno-unused-parameter`. This often reveals interesting warnings. In the case of aste:

```
~/github/aste/build [develop]$ make
[  6%] Building CXX object CMakeFiles/metisAPI.dir/src/metisAPI.cpp.o
/home/makish/github/aste/src/metisAPI.cpp: In function ‘void partitionMetis(idx_t, idx_t, idx_t*, idx_t*, idx_t, idx_t*)’:
/home/makish/github/aste/src/metisAPI.cpp:13:22: warning: unused variable ‘result’ [-Wunused-variable]
   13 |   int                result = METIS_PartMeshNodal(&cell_count, &point_count, cellptr, celldata, 0, 0, &nparts, 0, options, &objval, cell_partition.data(), point_partition);
      |                      ^~~~~~
[ 12%] Linking CXX shared library libmetisAPI.so
[ 12%] Built target metisAPI
[ 18%] Building CXX object CMakeFiles/precice-aste-run.dir/src/precice-aste-run.cpp.o
[ 25%] Building CXX object CMakeFiles/precice-aste-run.dir/src/common.cpp.o
[ 31%] Building CXX object CMakeFiles/precice-aste-run.dir/src/mesh.cpp.o
/home/makish/github/aste/src/mesh.cpp: In member function ‘void aste::MeshName::save(const aste::Mesh&, const string&) const’:
/home/makish/github/aste/src/mesh.cpp:324:26: warning: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘vtkIdType’ {aka ‘long long int’} [-Wsign-compare]
  324 |     for (size_t i = 0; i < grid->GetNumberOfPoints(); i++) {
      |                        ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 37%] Building CXX object CMakeFiles/precice-aste-run.dir/src/configreader.cpp.o
/home/makish/github/aste/src/configreader.cpp: In member function ‘void aste::asteConfig::load(const string&)’:
/home/makish/github/aste/src/configreader.cpp:12:30: warning: catching polymorphic type ‘class nlohmann::detail::parse_error’ by value [-Wcatch-value=]
   12 |   } catch (nlohmann::detail::parse_error) {
      |                              ^~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:15:30: warning: catching polymorphic type ‘class nlohmann::detail::type_error’ by value [-Wcatch-value=]
   15 |   } catch (nlohmann::detail::type_error) {
      |                              ^~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:22:30: warning: catching polymorphic type ‘class nlohmann::detail::parse_error’ by value [-Wcatch-value=]
   22 |   } catch (nlohmann::detail::parse_error) {
      |                              ^~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:25:30: warning: catching polymorphic type ‘class nlohmann::detail::type_error’ by value [-Wcatch-value=]
   25 |   } catch (nlohmann::detail::type_error) {
      |                              ^~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:32:30: warning: catching polymorphic type ‘class nlohmann::detail::type_error’ by value [-Wcatch-value=]
   32 |   } catch (nlohmann::detail::type_error) {
      |                              ^~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:35:32: warning: catching polymorphic type ‘class nlohmann::detail::type_error’ by value [-Wcatch-value=]
   35 |     } catch (nlohmann::detail::type_error) {
      |                                ^~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:38:19: warning: catching polymorphic type ‘class std::invalid_argument’ by value [-Wcatch-value=]
   38 |     } catch (std::invalid_argument) {
      |                   ^~~~~~~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:60:32: warning: catching polymorphic type ‘class nlohmann::detail::parse_error’ by value [-Wcatch-value=]
   60 |     } catch (nlohmann::detail::parse_error) {
      |                                ^~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:63:32: warning: catching polymorphic type ‘class nlohmann::detail::type_error’ by value [-Wcatch-value=]
   63 |     } catch (nlohmann::detail::type_error) {
      |                                ^~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:70:32: warning: catching polymorphic type ‘class nlohmann::detail::parse_error’ by value [-Wcatch-value=]
   70 |     } catch (nlohmann::detail::parse_error) {
      |                                ^~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:73:32: warning: catching polymorphic type ‘class nlohmann::detail::type_error’ by value [-Wcatch-value=]
   73 |     } catch (nlohmann::detail::type_error) {
      |                                ^~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:83:24: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const long unsigned int’ [-Wsign-compare]
   83 |     for (auto k = 0; k < readScalarSize; k++) {
      |                      ~~^~~~~~~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:87:24: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const long unsigned int’ [-Wsign-compare]
   87 |     for (auto k = 0; k < readVectorSize; k++) {
      |                      ~~^~~~~~~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:91:24: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const long unsigned int’ [-Wsign-compare]
   91 |     for (auto k = 0; k < writeScalarSize; k++) {
      |                      ~~^~~~~~~~~~~~~~~~~
/home/makish/github/aste/src/configreader.cpp:95:24: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const long unsigned int’ [-Wsign-compare]
   95 |     for (auto k = 0; k < writeVectorSize; k++) {
      |                      ~~^~~~~~~~~~~~~~~~~
[ 43%] Building CXX object CMakeFiles/precice-aste-run.dir/src/modes.cpp.o
/home/makish/github/aste/src/modes.cpp: In function ‘void aste::runReplayMode(const aste::ExecutionContext&, const string&)’:
/home/makish/github/aste/src/modes.cpp:116:56: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  116 |   while (preciceInterface.isCouplingOngoing() && round < minMeshSize) {
      |                                                  ~~~~~~^~~~~~~~~~~~~
[ 50%] Building CXX object CMakeFiles/precice-aste-run.dir/src/utilities.cpp.o
In file included from /usr/include/c++/11/cassert:44,
                 from /home/makish/github/aste/src/modes.hpp:3,
                 from /home/makish/github/aste/src/utilities.hpp:21,
                 from /home/makish/github/aste/src/utilities.cpp:1:
/home/makish/github/aste/src/utilities.cpp: In function ‘std::vector<int, std::allocator<int> > aste::setupVertexIDs(precice::SolverInterface&, const aste::Mesh&, int)’:
/home/makish/github/aste/src/utilities.cpp:44:23: warning: comparison of integer expressions of different signedness: ‘std::vector<double>::size_type’ {aka ‘long unsigned int’} and ‘const int’ [-Wsign-compare]
   44 |     assert(pos.size() == dimension);
      |            ~~~~~~~~~~~^~~~~~~~~~~~
[ 56%] Building CXX object CMakeFiles/precice-aste-run.dir/src/logger.cpp.o
[ 62%] Linking CXX executable precice-aste-run
[ 62%] Built target precice-aste-run
[ 68%] Building CXX object CMakeFiles/testing.dir/tests/testing.cpp.o
[ 75%] Building CXX object CMakeFiles/testing.dir/tests/read_test.cpp.o
[ 81%] Building CXX object CMakeFiles/testing.dir/tests/write_test.cpp.o
[ 87%] Building CXX object CMakeFiles/testing.dir/src/mesh.cpp.o
/home/makish/github/aste/src/mesh.cpp: In member function ‘void aste::MeshName::save(const aste::Mesh&, const string&) const’:
/home/makish/github/aste/src/mesh.cpp:324:26: warning: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘vtkIdType’ {aka ‘long long int’} [-Wsign-compare]
  324 |     for (size_t i = 0; i < grid->GetNumberOfPoints(); i++) {
      |                        ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 93%] Building CXX object CMakeFiles/testing.dir/src/logger.cpp.o
[100%] Linking CXX executable testing
[100%] Built target testing
```

--- a/src/configreader.cpp
+++ b/src/configreader.cpp
@@ -9,33 +9,33 @@ void asteConfig::load(const std::string &asteConfigFile)
 
   try {
     preciceConfigFilename = config["precice-config"].get<std::string>();
-  } catch (nlohmann::detail::parse_error) {
+  } catch (nlohmann::detail::parse_error &) {
     ASTE_ERROR << "Error while parsing ASTE configuration file \"precice-config\" is missing";
     MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
-  } catch (nlohmann::detail::type_error) {
+  } catch (nlohmann::detail::type_error &) {
     ASTE_ERROR << "Error while parsing ASTE configuration file \"precice-config\" is missing";
     MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
   }
 
   try {
     participantName = config["participant"].get<std::string>();
-  } catch (nlohmann::detail::parse_error) {
+  } catch (nlohmann::detail::parse_error &) {
     ASTE_ERROR << "Error while parsing ASTE configuration file \"participant\" is missing";
     MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
-  } catch (nlohmann::detail::type_error) {
+  } catch (nlohmann::detail::type_error &) {
     ASTE_ERROR << "Error while parsing ASTE configuration file \"participant\" is missing";
     MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
   }
 
   try {
     startdt = config["startdt"].get<int>();
-  } catch (nlohmann::detail::type_error) {
+  } catch (nlohmann::detail::type_error &) {
     try {
       startdt = std::stoi(config["startdt"].get<std::string>());
-    } catch (nlohmann::detail::type_error) {
+    } catch (nlohmann::detail::type_error &) {
       ASTE_ERROR << "Error while parsing ASTE configuration file \"startdt\" is missing or has a wrong type, it must be an integer or integer convertible string.";
       MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
-    } catch (std::invalid_argument) {
+    } catch (std::invalid_argument &) {
       ASTE_ERROR << "Error while parsing startdt from ASTE configuration file it must be an integer or integer convertible string.";
       MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
     }
@@ -57,20 +57,20 @@ void asteConfig::load(const std::string &asteConfigFile)
     asteInterface interface;
     try {
       interface.meshName = config["meshes"][i]["mesh"].get<std::string>();
-    } catch (nlohmann::detail::parse_error) {
+    } catch (nlohmann::detail::parse_error &) {
       ASTE_ERROR << "Error while parsing ASTE configuration file \"mesh\" is missing";
       MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
-    } catch (nlohmann::detail::type_error) {
+    } catch (nlohmann::detail::type_error &) {
       ASTE_ERROR << "Error while parsing ASTE configuration file \"mesh\" is missing or not a string";
       MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
     }
 
     try {
       interface.meshFilePrefix = config["meshes"][i]["meshfileprefix"];
-    } catch (nlohmann::detail::parse_error) {
+    } catch (nlohmann::detail::parse_error &) {
       ASTE_ERROR << "Error while parsing ASTE configuration file \"meshfileprefix\" is missing";
       MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
-    } catch (nlohmann::detail::type_error) {
+    } catch (nlohmann::detail::type_error &) {
       ASTE_ERROR << "Error while parsing ASTE configuration file \"meshfileprefix\" is missing or not a string";
       MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
     }
@@ -80,19 +80,19 @@ void asteConfig::load(const std::string &asteConfigFile)
     const auto writeScalarSize = config["meshes"][i]["write-data"]["scalar"].size();
     const auto writeVectorSize = config["meshes"][i]["write-data"]["vector"].size();
 
-    for (auto k = 0; k < readScalarSize; k++) {
+    for (std::size_t k = 0; k < readScalarSize; k++) {
       const auto scalarName = config["meshes"][i]["read-data"]["scalar"][k];
       interface.readScalarNames.push_back(scalarName);
     }
-    for (auto k = 0; k < readVectorSize; k++) {
+    for (std::size_t k = 0; k < readVectorSize; k++) {
       const auto vectorName = config["meshes"][i]["read-data"]["vector"][k];
       interface.readVectorNames.push_back(vectorName);
     }
-    for (auto k = 0; k < writeScalarSize; k++) {
+    for (std::size_t k = 0; k < writeScalarSize; k++) {
       const auto scalarName = config["meshes"][i]["write-data"]["scalar"][k];
       interface.writeScalarNames.push_back(scalarName);
     }
-    for (auto k = 0; k < writeVectorSize; k++) {
+    for (std::size_t k = 0; k < writeVectorSize; k++) {
       const auto vectorName = config["meshes"][i]["write-data"]["vector"][k];
       interface.writeVectorNames.push_back(vectorName);
     }

--- a/src/mesh.cpp
+++ b/src/mesh.cpp
@@ -321,7 +321,7 @@ void MeshName::save(const Mesh &mesh, const std::string &outputFileName) const
 
     std::vector<double> pointData;
     pointData.reserve(3);
-    for (size_t i = 0; i < grid->GetNumberOfPoints(); i++) {
+    for (vtkIdType i = 0; i < grid->GetNumberOfPoints(); i++) {
       for (int j = 0; j < meshdata.numcomp; j++) {
         pointData.push_back(meshdata.dataVector[i * meshdata.numcomp + j]);
       }

--- a/src/metisAPI.cpp
+++ b/src/metisAPI.cpp
@@ -10,7 +10,8 @@ void partitionMetis(idx_t cell_count, idx_t point_count, idx_t *cellptr, idx_t *
   METIS_SetDefaultOptions(options);
   std::vector<idx_t> cell_partition(cell_count);
   idx_t              objval;
-  int                result = METIS_PartMeshNodal(&cell_count, &point_count, cellptr, celldata, 0, 0, &nparts, 0, options, &objval, cell_partition.data(), point_partition);
+  // TODO: Check return value of the function (and potentially add an assert)
+  METIS_PartMeshNodal(&cell_count, &point_count, cellptr, celldata, 0, 0, &nparts, 0, options, &objval, cell_partition.data(), point_partition);
 }
 
 int typewidth()

--- a/src/modes.cpp
+++ b/src/modes.cpp
@@ -68,7 +68,7 @@ void aste::runReplayMode(const aste::ExecutionContext &context, const std::strin
   }
 
   dt = preciceInterface.initialize();
-  int round{0};
+  std::size_t round{0};
 
   ASTE_DEBUG << "Looking for dt = " << asteConfiguration.startdt;
   for (const auto &mesh : asteConfiguration.asteInterfaces.front().meshes) {
@@ -113,7 +113,7 @@ void aste::runReplayMode(const aste::ExecutionContext &context, const std::strin
   const std::string &coric = precice::constants::actionReadIterationCheckpoint();
   const std::string &cowic = precice::constants::actionWriteIterationCheckpoint();
 
-  while (preciceInterface.isCouplingOngoing() && round < minMeshSize) {
+  while (preciceInterface.isCouplingOngoing() && (round < minMeshSize)) {
     if (preciceInterface.isActionRequired(cowic)) {
       ASTE_ERROR << "Implicit coupling schemes cannot be used with ASTE";
       MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);

--- a/src/utilities.cpp
+++ b/src/utilities.cpp
@@ -41,7 +41,7 @@ std::vector<int> aste::setupVertexIDs(precice::SolverInterface &interface,
   std::vector<double> posData(dimension * nvertices);
   for (unsigned long i = 0; i < nvertices; ++i) {
     const auto &pos = mesh.positions[i];
-    assert(pos.size() == dimension);
+    assert(pos.size() == static_cast<unsigned int>(dimension));
     std::copy(pos.begin(), pos.end(), &posData[i * dimension]);
   }
 

